libnopegl/src/glsl/helper_noise.glsl

```diff
@@ -0,0 +1,161 @@
+ */
+highp uint lowbias32(highp uint x)
+{
+    x ^= x >> 16;
+    x *= 0x7feb352dU;
+    x ^= x >> 15;
+    x *= 0x846ca68bU;
+    x ^= x >> 16;
+    return x;
+}
+
+uint hash(uvec2 x)
+{
+    return lowbias32(x.x ^ lowbias32(x.y));
+}
+
+uint hash(uvec3 x)
+{
+    return lowbias32(x.x ^ lowbias32(x.y ^ lowbias32(x.z)));
+}
+
+float random(uvec3 x)
+{
+    return u32tof32(hash(x));
+}
+
+/*
+ * Generate a random point on a sphere (with uniform distribution)
+ * 
+ * References:
+ * - https://web.archive.org/web/20200618040237/https://mrl.nyu.edu/~perlin/paper445.pdf
+ * - https://web.archive.org/web/20180616011332/http://www.scratchapixel.com/lessons/procedural-generation-virtual-worlds%20/perlin-noise-part-2
+ */
+vec3 random_grad(uvec3 x)
+{
+    /* Generate 2 random floats in [0;1] using 16-bit of information for each */
+    uint h = hash(x);
+    float r0 = u32tof32(0x00010001U * (h      & 0xffffU));
+    float r1 = u32tof32(0x00010001U * (h>>16  & 0xffffU));
+    float theta = acos(2.0 * r0 - 1.0); // use the first random for the polar angle (latitude)
+    float phi = ngli_tau * r1; // use the 2nd random for the azimuth (longitude)
+    return vec3(cos(phi) * sin(theta), sin(phi) * sin(theta), cos(theta));
+}
+
+vec3 curve_quintic(vec3 t)
+{
+    return ((6. * t - 15.) * t + 10.) * t * t * t;
+}
+
+float noise_perlin(vec3 t, uint seed)
+{
+    vec3 i = floor(t);
+    vec3 f = fract(t);
+    uvec3 x = uvec3(i) + seed;
+
+    float y0 = dot(random_grad(x + uvec3(0, 0, 0)), f - vec3(0, 0, 0));
+    float y1 = dot(random_grad(x + uvec3(1, 0, 0)), f - vec3(1, 0, 0));
+    float y2 = dot(random_grad(x + uvec3(0, 1, 0)), f - vec3(0, 1, 0));
+    float y3 = dot(random_grad(x + uvec3(1, 1, 0)), f - vec3(1, 1, 0));
+    float y4 = dot(random_grad(x + uvec3(0, 0, 1)), f - vec3(0, 0, 1));
+    float y5 = dot(random_grad(x + uvec3(1, 0, 1)), f - vec3(1, 0, 1));
+    float y6 = dot(random_grad(x + uvec3(0, 1, 1)), f - vec3(0, 1, 1));
+    float y7 = dot(random_grad(x + uvec3(1, 1, 1)), f - vec3(1, 1, 1));
+
+    vec3 a = curve_quintic(f);
+    return mix(mix(mix(y0, y1, a.x), mix(y2, y3, a.x), a.y),
+               mix(mix(y4, y5, a.x), mix(y6, y7, a.x), a.y),
+               a.z);
+}
+
+/* Adapted from https://en.wikipedia.org/wiki/Bicubic_interpolation */
+float bicubic(float a, float b, float c, float d, float t)
+{
+    mat4 m = 0.5 * mat4(
+        0.,  2.,  0.,  0.,
+       -1.,  0.,  1.,  0.,
+        2., -5.,  4., -1.,
+       -1.,  3., -3.,  1.
+    );
+    return dot(m * vec4(1., t, t*t, t*t*t), vec4(a, b, c, d));
+}
+
+float noise_blocky(vec3 t, uint seed)
+{
+    uvec2 xy = uvec2(t.xy) + seed;
+
+    /*
+     * For each latice coordinates (x, y) determine a random frequency and
+     * offset that will be used to compute the evolution
+     */
+    float evolution_freq = random(uvec3(xy + 1u << 16, seed));
+    float evolution_offset = random(uvec3(xy + 1u << 16, seed) << 16U);
```

Comment:
Would it be possible to use the same technique used in `random_grad()`?

```glsl
vec2 random2(uvec2 p)
{
    uint h = hash(p);
    float r0 = u32tof32(0x00010001U * (h      & 0xffffU));
    float r1 = u32tof32(0x00010001U * (h>>16  & 0xffffU));
    return vec2(r0, r1);
}

/* ... */

    r = random2(xy);
    float evolution_freq = r.x;
    float eolution_offset = r.y;
```

?

Reply:
Done.